pkg/server/sandbox_portforward_unix.go

```diff
@@ -62,48 +61,56 @@ func (c *criService) portForward(ctx context.Context, id string, port int32, str
 		netNSPath = "host"
 	}
 
-	socat, err := exec.LookPath("socat")
-	if err != nil {
-		return errors.Wrap(err, "failed to find socat")
-	}
-
-	// Check https://linux.die.net/man/1/socat for meaning of the options.
-	args := []string{socat, "-", fmt.Sprintf("TCP4:localhost:%d", port)}
-
-	log.G(ctx).Infof("Executing port forwarding command %q in network namespace %q", strings.Join(args, " "), netNSPath)
+	log.G(ctx).Infof("Executing port forwarding in network namespace %q", netNSPath)
 	err = netNSDo(func(_ ns.NetNS) error {
-		cmd := exec.Command(args[0], args[1:]...)
-		cmd.Stdout = stream
-
-		stderr := new(bytes.Buffer)
-		cmd.Stderr = stderr
-
-		// If we use Stdin, command.Run() won't return until the goroutine that's copying
-		// from stream finishes. Unfortunately, if you have a client like telnet connected
-		// via port forwarding, as long as the user's telnet client is connected to the user's
-		// local listener that port forwarding sets up, the telnet session never exits. This
-		// means that even if socat has finished running, command.Run() won't ever return
-		// (because the client still has the connection and stream open).
-		//
-		// The work around is to use StdinPipe(), as Wait() (called by Run()) closes the pipe
-		// when the command (socat) exits.
-		in, err := cmd.StdinPipe()
+		// TODO: hardcoded to tcp4 because localhost resolves to ::1 by default if the system has IPv6 enabled.
+		// Theoretically happy eyeballs will try IPv6 first and fallback to IPv4
+		// but resolving localhost doesn't seem to return and IPv4 address, thus failing the connection.
+		conn, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))
 		if err != nil {
-			return errors.Wrap(err, "failed to create stdin pipe")
+			return errors.Wrapf(err, "failed to dial %d", port)
 		}
+
+		errCh := make(chan error, 2)
+		// Copy from the client stream to the namespace port connection
 		go func() {
-			if _, err := io.Copy(in, stream); err != nil {
-				logrus.WithError(err).Errorf("Failed to copy port forward input for %q port %d", id, port)
-			}
-			in.Close()
-			logrus.Debugf("Finish copying port forward input for %q port %d", id, port)
+			// tries to shutdown only its write stream, giving it a chance to terminate gracefully
+			defer conn.Close()
```

Comment:
did you want to defer the close till after the timeout?

Reply:
One of the problems I've observed is that despite it forwards the port, once the client disconnect it was leaking connections inside the namespace. I think this guarantees that we always close the connection.
Is it possible to not leak connections without deferring?

Reply:
my question is to the desired scope of the deferred close.. Did you want to defer the close to the end of each go func?  If just for connection leaks I'd think about doing these defers at the `func(_ ns.NetNS) error { ` scope.  We probably need to be using cancellable context and check for case <-ctx.Done(): to make sure we get a chance to close. Hate connection leaks :-)

Reply:
ahh ok, it has to happen inside the go func() because it "closes" the other direction and "gives a chance to terminate gracefully", otherwise we'll always end with the timeout.
I think that this way we are mimicking socat completely

Reply:
I thought it was leaking before, not sure we should mimic that :)  Will this way not leak on context close? Closing the other direction abruptly ends the other direction with no bytes and EOF.. which is not as graceful as a select with wait for context close or timeout then close forcefully.  If my read of the code is right.. it can't timeout unless the close is locked out for some reason?  Did socat always time out?  hmm

Reply:
> mimicking socat completely

I mean socat itself, not to  the previous implementation with socat :sweat_smile: 

I think that this comment in the code we are replacing is significant too, using the pipe it was forcing to close everything
 
https://github.com/containerd/cri/pull/1470/files#diff-a60e53338184ff4aee2ff77caed7cabaL81-L89

> 		// If we use Stdin, command.Run() won't return until the goroutine that's copying	
> 		// from stream finishes. Unfortunately, if you have a client like telnet connected	
> 		// via port forwarding, as long as the user's telnet client is connected to the user's	
> 		// local listener that port forwarding sets up, the telnet session never exits. This	
> 		// means that even if socat has finished running, command.Run() won't ever return	
> 		// (because the client still has the connection and stream open).	
> 		//	
> 		// The work around is to use StdinPipe(), as Wait() (called by Run()) closes the pipe	
> 		// when the command (socat) exits



> Closing the other direction abruptly ends the other direction with no bytes and EOF..

we are closing the writer, so `io.Copy` will only exit if it tries to write, or the consumer of that stream close the connection (gracefully), otherwise, it will keep waiting forever ... that's the problem described by Lantao here https://github.com/containerd/cri/pull/729#issue-180773387

The previous attempt to replace socat was using a waiting group, hence the deadlock, currently we only need any of the streams to close and then we try to close everything, giving an opportunity to the applications that are using the stream to close it by themselves.